RESTserver/src/main/java/com/softserve/if072/restservice/service/ForecastService.java

```diff
@@ -0,0 +1,316 @@
+        double upperThreeSigmaLimit = productStatistics.getUpperThreeSigmaLimit();
+        boolean isOutlierAtTheBeginningOrEnd = false;
+        int leftTrimBound = 0;
+        int rightTrimBound = productUsedSpeeds.length;
+
+        /*
+        * Find outliers at the beginning of the data set
+        */
+        for (int i = 0; i < productUsedSpeeds.length; i++) {
+            if (productUsedSpeeds[i] < lowerThreeSigmaLimit || productUsedSpeeds[i] > upperThreeSigmaLimit) {
+                if (i == 0) {
+                    isOutlierAtTheBeginningOrEnd = true;
+                    leftTrimBound++;
+                    continue;
+                }
+                leftTrimBound++;
+                continue;
+            }
+            LOGGER.trace("Find {} outlier(s) at the beginning of the data set.", i);
+            break;
+        }
+
+        /*
+        * Find outliers at the end of the data set
+        */
+        for (int i = productUsedSpeeds.length - 1; i >= 0; i--) {
+            if (productUsedSpeeds[i] < lowerThreeSigmaLimit || productUsedSpeeds[i] > upperThreeSigmaLimit) {
+                if (i == productUsedSpeeds.length - 1) {
+                    isOutlierAtTheBeginningOrEnd = true;
+                    rightTrimBound--;
+                    continue;
+                }
+                rightTrimBound--;
+                continue;
+            }
+            LOGGER.trace("Find {} outlier(s) at the end of the data set.", productUsedSpeeds.length - 1 - i);
+            break;
+        }
+
+        /*
+        * Trim outliers from the beginning and the end of the data set
+        */
+        if (isOutlierAtTheBeginningOrEnd) {
+            hasOutliers = true;
+            productUsedSpeeds = Arrays.copyOfRange(productUsedSpeeds, leftTrimBound, rightTrimBound);
+        }
+
+        /*
+        * Find and deal with other outliers
+        */
+        int otherOutliersCount = 0;
+        for (int i = 1; i < productUsedSpeeds.length - 1; i++) {
+            if (productUsedSpeeds[i] < lowerThreeSigmaLimit || productUsedSpeeds[i] > upperThreeSigmaLimit) {
+                hasOutliers = true;
+                otherOutliersCount++;
+                int lastOutlierIndex = i;
+                while (true) {
+                    if (!(productUsedSpeeds[lastOutlierIndex + 1] < lowerThreeSigmaLimit || productUsedSpeeds[lastOutlierIndex + 1] > upperThreeSigmaLimit)) {
+                        break;
+                    }
+                    otherOutliersCount++;
+                    lastOutlierIndex++;
+                }
+                for (int k = i; k <= lastOutlierIndex; k++) {
+                    productUsedSpeeds[k] = (productUsedSpeeds[i - 1] + productUsedSpeeds[lastOutlierIndex + 1]) / 2;
+                }
+                LOGGER.trace("Find outlier(s) from the index {} to the index {}.", i, lastOutlierIndex);
+                i = lastOutlierIndex + 1;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor the code in order to not assign to this loop counter from within the loop body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AForLoopCounterChangedCheck)